src/driver/franka_plan_runner.cc

```diff
@@ -763,45 +781,65 @@ franka::JointPositions FrankaPlanRunner::JointPositionCallback(
-        "starting initial timestep...",
-        plan_utime_);
-    // first time step of plan, reset time and start conf
-    franka_time_ = 0.0;
-    start_conf_plan_ = plan_->value(franka_time_);
-
-    if (!LimitJoints(start_conf_plan_)) {
-      dexai::log()->warn(
-          "JointPositionCallback: plan {} at franka_time_: {} seconds "
-          "is exceeding the joint limits!",
-          plan_utime_, franka_time_);
-    }
 
-    // the current (desired) position of franka is the starting position:
-    start_conf_franka_ = current_conf_franka;
-    end_conf_plan_ = plan_->value(plan_->end_time());
+    auto [new_plan, new_plan_utime, new_plan_exec_opt,
+          new_plan_contact_expected] {comm_interface_->PopNewPlan()};
 
-    // Maximum change in joint angle between two confs
-    auto max_ang_distance =
-        utils::max_angular_distance(start_conf_franka_, start_conf_plan_);
-    if (max_ang_distance > params_.kMediumJointDistance) {
-      dexai::log()->error(
-          "JointPositionCallback: Discarding plan, mismatched start position."
-          " Max distance: {} > {}",
-          max_ang_distance, params_.kMediumJointDistance);
-      comm_interface_->PublishPlanComplete(
-          plan_utime_, false, "discarded due to mismatched start conf");
-      ResetPlan();
-      return franka::MotionFinished(franka::JointPositions(robot_state.q));
-    } else if (max_ang_distance > params_.kTightJointDistance) {
-      dexai::log()->warn(
-          "JointPositionCallback: max angular distance between franka and "
-          "start of plan is larger than 'kTightJointDistance': {} > {}",
-          max_ang_distance, params_.kTightJointDistance);
+    bool continue_from_current_plan {plan_ != nullptr && !new_plan->empty()
+                                     && new_plan->end_time() > franka_time_};
+    if (!continue_from_current_plan || IsContinuous(new_plan)) {
+      plan_ = std::move(new_plan);
+      plan_utime_ = new_plan_utime;
+      plan_exec_opt_ = new_plan_exec_opt;
+      contact_expected_ = new_plan_contact_expected;
+
+      plan_start_utime_ = utils::get_current_utime();
+
+      dexai::log()->info(
+          "JointPositionCallback: popped new plan {} from buffer, "
+          "starting initial timestep...",
+          plan_utime_);
+
+      if (!continue_from_current_plan) {
+        // first time step of plan, reset time and start conf
+        franka_time_ = 0.0;
+      }
+
+      start_conf_plan_ = plan_->value(franka_time_);
+
+      if (!LimitJoints(start_conf_plan_)) {
+        dexai::log()->warn(
+            "JointPositionCallback: plan {} at franka_time_: {} seconds "
+            "is exceeding the joint limits!",
+            plan_utime_, franka_time_);
+      }
+
+      // the current (desired) position of franka is the starting position:
+      start_conf_franka_ = current_conf_franka;
+      end_conf_plan_ = plan_->value(plan_->end_time());
+    }
```

Comment:
does it make sense to package this code as a separate method to enhance readability? `OverwritePlan`? `ResetPlan`? `UpdatePlan`?

Reply:
What does `ResetPlan` already do?

Reply:
`ResetPlan` clears the `plan_`, resets `franka_time_`, etc.

Reply:
I have wrapped the printing in a function, but the problem with packaging the rest in a function is that we need `has_active_plan` anyways, and `plan_` is a `unique_ptr`

Reply:
Done